sdks/python/apache_beam/runners/worker/operations.py

```diff
@@ -157,6 +159,24 @@ def add_receiver(self, operation, output_index=0):
     """Adds a receiver operation for the specified output."""
     self.consumers[output_index].append(operation)
 
+  def progress_metrics(self):
+    return beam_fn_api_pb2.Metrics.PTransform(
+        processed_elements=beam_fn_api_pb2.Metrics.PTransform.ProcessedElements(
+            measured=beam_fn_api_pb2.Metrics.PTransform.Measured(
+                total_time_spent=(
```

Comment:
Do you think that we should have specific blocks in Metrics.PTransform which measure StartBundle/FinishBundle explicitly and not add them here as part of the ProcessedElements/ActiveElements section or should those fall under different areas?

My concern is that we would want to add information about how expensive every logical unit of work is like splitting an SDF and we would have metrics for every SDK and how it has chosen to represent a DoFn and its logical portions.

Reply:
I think we should keep it simple for now. 

We've talked about this many times in the Autoscaling team (and I had discussions recently about this as well). Start/Finish aren't as straightforward as one might think, e.g. often there's some one-time initialization based on side inputs (this happens in all the ML pipelines) and the work of FinishBundle can be triggered on arbitrary elements too (e.g. batching) and often have cost proportional to the number of input elements rather than fixed. I think our systems need to be robust to such lumpiness. There's Setup and Teardown that we may want to profile as well. 

We do know that these basic measurements are important and can extend this to have more signals if needed in the future. There are other possible extensions too (e.g. serialized element size has historically often represented work better than element count, and this could be generalized to element weights which could be particularly useful for SDFs).

Reply:
sgtm